Mathlib/Topology/CWComplex.lean

```diff
@@ -0,0 +1,123 @@
+/-
+Copyright (c) 2024 Elliot Dean Young and Jiazhen Xia. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jiazhen Xia, Elliot Dean Young
+-/
+
+import Mathlib.Topology.Category.TopCat.Limits.Basic
+import Mathlib.CategoryTheory.Functor.OfSequence
+
+/-!
+# CW-complexes
+
+This file defines (relative) CW-complexes.
+
+## Main definitions
+
+* `RelativeCWComplex`: A relative CW-complex is the colimit of an expanding sequence of subspaces
+`sk i` (called the `(i-1)`-skeleton) for `i ≥ 0`, where `sk 0` (i.e., the `(-1)`-skeleton) is an
+arbitrary topological space, and each `sk (n+1)` (i.e., the `n`-skeleton) is obtained from `sk n`
+(i.e., the `(n-1)`-skeleton) by attaching `n`-disks.
+
+* `CWComplex`: A CW-complex is a relative CW-complex whose `sk 0` (i.e., `(-1)`-skeleton) is empty.
+
+## References
+
+The definition of CW-complexes follows David Wärn's suggestion at
+https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/Do.20we.20have.20CW.20complexes.3F/near/231769080
+-/
+
+open CategoryTheory TopCat
+
+universe u
+
+namespace RelativeCWComplex
+
+/-- The inclusion map from the `n`-sphere to the `(n+1)`-disk -/
+def sphereInclusion (n : ℤ) : 𝕊 n ⟶ 𝔻 (n + 1) where
+  toFun := fun ⟨p, hp⟩ ↦ ⟨p, le_of_eq hp⟩
+  continuous_toFun := ⟨fun t ⟨s, ⟨r, hro, hrs⟩, hst⟩ ↦ by
+    rw [isOpen_induced_iff, ← hst, ← hrs]
+    tauto⟩
+
+variable {S D : ℤ → TopCat.{u}} (f : ∀ n, S n ⟶ D (n + 1))
+
+/-- The inclusion map from the disjoint union of `S n` (boundary of generalized `(n+1)`-cells) to
+the disjoint union of `D (n + 1)` (generalized `(n+1)`-cells) where both of the disjoint unions are
+indexed by `cells` -/
+def generalizedSigmaSphereInclusion (n : ℤ) (cells : Type) :
+    TopCat.of (Σ (_ : cells), S n) ⟶ TopCat.of (Σ (_ : cells), D (n + 1)) where
+  toFun := Sigma.map id fun _ x ↦ (f n).toFun x
+  continuous_toFun := Continuous.sigma_map fun _ ↦ (f n).continuous_toFun
+
+/-- Given an attaching map for each `S n` (boundary of a generalized `(n+1)`-cell), we construct
+the attaching map for the disjoint union of all the `S n`. -/
+def generalizedSigmaAttachMap (X : TopCat.{u}) (n : ℤ) (cells : Type)
+    (attach_maps : cells → C(S n, X)) : TopCat.of (Σ (_ : cells), S n) ⟶ X where
+  toFun := fun ⟨i, x⟩ ↦ attach_maps i x
+  continuous_toFun := continuous_sigma fun i ↦ (attach_maps i).continuous_toFun
+
+/-- A type witnessing that `X'` is obtained from `X` by attaching generalized `(n+1)`-cells, where
+a generalized `(n+1)`-cell is given by `f n : S n ⟶ D (n + 1)`. -/
+structure AttachGeneralizedCells (X X' : TopCat.{u}) (n : ℤ) where
```

Comment:
Could you make this depend only on a *single* map `f` between topological spaces rather than a sequence of maps and an index `n`? It should not require too many changes to your code, but it would ease the future development of the API for `AttachGeneralizedCells`.